mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp

```diff
@@ -1097,6 +1098,440 @@ MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+    return failure();
+  }
+
+  FailureOr<llvm::SmallDenseMap<int64_t, SmallVector<SubDimInfo>>>
+      lowerSubDims =
+          getLowerSubDimensions(rewriter, toBeReducedViews, {0, 1, 2});
+  if (failed(lowerSubDims)) {
+    LLVM_DEBUG(llvm::dbgs() << "lowerSubDims creation using "
+                               "getLowerSubDimensions is unsuccesful.\n");
+    return failure();
+  }
+
+  int64_t reductionAxis = reduceOp.getAxisAttr().getInt();
+  TypedValue<ShapedType> redOut = reduceOp.getOut();
+  ArrayRef<int64_t> reduceOutShape = redOut.getType().getShape();
+  TypedValue<ShapedType> redIn = reduceOp.getIn();
+  ArrayRef<int64_t> reduceInShape = redIn.getType().getShape();
+
+  int64_t blockReductionAxis = reductionAxis;
+  int64_t blockReductionAxisFromLeft =
+      (reduceInShape.size() - 1) - blockReductionAxis;
+
+  ArrayRef<int64_t> blockLowerShape = getLowerShape(blockSubTileViews.value());
+  ArrayRef<int64_t> blockSubTileTidSliceShape =
+      getLowerShape(blockSubTileTidSliceViews.value());
+  int64_t blockSubTileTidSliceRank = blockSubTileTidSliceShape.size();
+  // The block sub-tile view might not have the slower changing
+  // dimensions in it. Thus, we always keep track of the reduction
+  // dimensions from its distance to fastest changing dimensions.
+  blockReductionAxis =
+      blockSubTileTidSliceRank - 1 - blockReductionAxisFromLeft;
+  int64_t partialReductionsPerThread =
+      blockSubTileTidSliceShape[blockReductionAxis];
+  int64_t ldsWorkspaceSize = 1;
+  for (auto [idx, size] : llvm::enumerate(blockLowerShape)) {
+    if (idx == (size_t)blockReductionAxis) {
+      ldsWorkspaceSize *= partialReductionsPerThread;
+    } else {
+      ldsWorkspaceSize *= size;
+    }
+  }
+  TypedValue<MemRefType> src = threadwiseWriteOp.getSource();
+  auto broadcastReducedSrc = rewriter.create<GpuAllocOp>(loc, src.getType());
+  Value ldsWorkspace = rock::gpuAlloc(rewriter, loc, ldsWorkspaceSize,
+                                      src.getType().getElementType(),
+                                      gpu::AddressSpace::Workgroup);
+
+  rewriter.create<BlockwiseBroadcastReduceOp>(
+      loc, src, ldsWorkspace, broadcastReducedSrc,
+      /*extraOut=*/nullptr, rewriter.getIndexAttr(blockReductionAxis),
+      reduceOp.getReduceMethodAttr(), blockSubTileViews.value(),
+      blockSubTileTidSliceViews.value(), threadSubTileViews.value(),
+      /*extraViews=*/nullptr,
+      getBlockSize(reduceOp->getParentOfType<func::FuncOp>()).value());
+
+  ViewLikeOpInterface viewOp =
+      ldsWorkspace.getDefiningOp<ViewLikeOpInterface>();
+  rewriter.create<GpuDeallocOp>(loc, viewOp.getViewSource());
+  // Create partial reduction views
+  ArrayAttr paddedReducedTrStack;
+  {
+    SmallVector<Attribute> transformAttrs;
+    ArrayRef<int64_t> blockTileShape = getLowerShape(blockSubTileViews.value());
+    SmallVector<SmallString<8>> names;
+    SmallVector<StringRef> nameRefs;
+    for (size_t i = 0; i < blockTileShape.size(); i++) {
```

Comment:
nit: llvm::enumerate()

Reply:
I ll remove the whole loop as I ve created a utility to do this now.